### PR TITLE
Pre filtering option for SRSNV

### PR DESCRIPTION
**Pull request by kenissur**
Added a pre_filter option to prepare_featuremap_for_model
Added pre_filter and random_seed to CLI

### CODE REVIEW

File: test/unit/mrd/test_srsnv_train_utils.py [hunk 1]
```diff
@@ -40,6 +41,40 @@ def test_prepare_featuremap_for_model(tmpdir):
...
+        train_set_size=100,
+        test_set_size=100,
+        balanced_sampling_info_fields=None,
+        pre_filter_bcftools_include=pre_filter_bcftools_include,
+        random_seed=0,
+    )
+    # In this scenario we are pre-filtering the test data so that only 4 FeatureMap entries pass:
+    total_variants = int(
+        subprocess.check_output(
+            f"bcftools view -H {input_featuremap_vcf} -i '{pre_filter_bcftools_include}' | wc -l",
```
Comment (doron-st) [inline test/unit/mrd/test_srsnv_train_utils.py:64]: can we have the pre_filter as a python function which takes the filter as argument, and this unit-test could test that the function works?

File: test/unit/mrd/test_srsnv_train_utils.py [hunk 2]
```diff
@@ -40,6 +41,40 @@ def test_prepare_featuremap_for_model(tmpdir):
...
+        )
+        .decode()
+        .strip()
+    )
+    assert total_variants == 4
+    # and since we are asking for more entries than are available, we should get all of them in equal ratios
+    n_variants = __count_variants(downsampled_training_featuremap_vcf)
+    assert n_variants == 2
+    n_variants = __count_variants(downsampled_test_featuremap_vcf)
+    assert n_variants == 2
```
Comment (doron-st) [inline test/unit/mrd/test_srsnv_train_utils.py:75]: lovely

File: ugvc/mrd/srsnv_inference_utils.py [hunk 3]
```diff
@@ -85,28 +143,105 @@ def process_records(self, records: list[pysam.VariantRecord]) -> list[pysam.Vari
...
-        predicted_qualities = self.model.predict_proba(df)
+        predicted_probability = self.model.predict_proba(df)
+        if self.pre_filter:
+            try:
+                # the bcftools filter should work as a query if applied on info fields because their names are the same
+                # in the vcf header and in the dataframe. However, the bcftools expression could explicitly contain
+                # INFO/, e.g INFO/X_SCORE>4, so we need to remove the INFO/ prefix before applying the filter
+                # We also replace && with & because the former is supported by bcftools and the latter by pandas
+                # We assign a value of 1 for error probability, translating to a quality of 0
+                predicted_probability[~df.eval(self.pre_filter.replace("INFO/", "").replace("&&", "&"))] = 1
```
Comment (doron-st) [inline ugvc/mrd/srsnv_inference_utils.py]: A tiny bit nicer to do the string processing in the constructor once.

File: ugvc/mrd/srsnv_inference_utils.py [hunk 4]
```diff
@@ -85,28 +143,105 @@ def process_records(self, records: list[pysam.VariantRecord]) -> list[pysam.Vari
...
+    Returns
+    -------
+    interp1d
+        Interpolation function
+    """
+    # read data, filter for ML_QUAL filters only
+    df = pd.read_parquet(mrd_simulation_dataframe)
+    df.index = df.index.str.upper()
+    df = df[df.index.str.startswith(ML_QUAL)]
+    df.loc[:, ML_QUAL] = df.index.str.replace(ML_QUAL + "_", "").astype(int)
```
Comment (doron-st) [inline ugvc/mrd/srsnv_inference_utils.py]: is rounding good for us here?

File: ugvc/mrd/srsnv_inference_utils.py [hunk 5]
```diff
@@ -85,28 +143,105 @@ def process_records(self, records: list[pysam.VariantRecord]) -> list[pysam.Vari
...
+    # Calculate Phred scores matching the residual SNV rate
+    phread_residual_snv_rate = -10 * np.log10(df["residual_snv_rate"])
+    # Create interpolation function
+    quality_interpolation_function = interp1d(
```
Comment (doron-st) [inline ugvc/mrd/srsnv_inference_utils.py:231]: cool

File: ugvc/mrd/srsnv_inference_utils.py [hunk 6]
```diff
@@ -85,28 +143,105 @@ def process_records(self, records: list[pysam.VariantRecord]) -> list[pysam.Vari
...
+
+    """
     df = pd.read_parquet(parquet_path)
     category_mappings = {}
     for column in df.select_dtypes(include=["category"]).columns:
         category_mappings[column] = df[column].cat.categories.tolist()
     return category_mappings
 
 
+def get_quality_interpolation_function(mrd_simulation_dataframe: str):
```
Comment (doron-st) [inline ugvc/mrd/srsnv_inference_utils.py]: can't mrd_simulation_dataframe be None?

File: ugvc/mrd/srsnv_plotting_utils.py [hunk 7]
```diff
@@ -249,44 +259,127 @@ def plot_ROC_curve(
         )
 
 
-def LoD_training_summary(
+def filter_valid_queries(df_test: pd.DataFrame, lod_filters: dict, verbose: bool = False) -> dict:
```
Comment (doron-st) [inline ugvc/mrd/srsnv_plotting_utils.py]: this method looks much too general o be in srsnv_plotting

File: ugvc/mrd/srsnv_plotting_utils.py [hunk 8]
```diff
@@ -249,44 +259,127 @@ def plot_ROC_curve(
...
+            df_test.eval(filter_query)
+            valid_filters[filter_name] = filter_query
+        except Exception:  # pylint: disable=broad-except
+            if verbose:
+                logger.warning(f"Filter query {filter_query} caused an exception, skipping.")
+
+    return valid_filters
+
+
+def count_bases_in_bed_file(file_path: str) -> int:
```
Comment (doron-st) [inline ugvc/mrd/srsnv_plotting_utils.py]: this one should be in vcfbed

**Review by doron-st — COMMENTED** [review]
Did my best. Very nice work as usual

File: ugvc/mrd/srsnv_inference_utils.py (hunk 6 above)
Comment (kenissur) [reply ugvc/mrd/srsnv_inference_utils.py]: Good catch, fixed

File: ugvc/mrd/srsnv_inference_utils.py (hunk 4 above)
Comment (kenissur) [reply ugvc/mrd/srsnv_inference_utils.py]: Again, good catch, fixed. Presently the dataframe contains integer values, encoded in the name, but in general no reason for it to be an integer.

File: ugvc/mrd/srsnv_inference_utils.py (hunk 3 above)
Comment (kenissur) [reply ugvc/mrd/srsnv_inference_utils.py]: Agree, done.

File: ugvc/mrd/srsnv_plotting_utils.py (hunk 8 above)
Comment (kenissur) [reply ugvc/mrd/srsnv_plotting_utils.py]: Moving

File: ugvc/mrd/srsnv_plotting_utils.py (hunk 7 above)
Comment (kenissur) [reply ugvc/mrd/srsnv_plotting_utils.py]: Moved to ugvc/utils/misc_utils.py
Makes sense?

**Review by ilyasoifer — APPROVED** [review]
Looks good, but I don't understand much

**Pull request merged**